### PR TITLE
Rollback partial kapacitor fails

### PR DESCRIPTION
**Pull request by itzg**
# Resolves

https://jira.rax.io/browse/SALUS-414

# What

If one or more of the discovered kapacitor instances successful accepted a task creation, but one of them failed we were leaving the successful ones configured with the kapacitor task. The entity wouldn't be saved, so there was no way to cleanup the partial configuration.

# How

This change implements a pseudo-transaction-rollback behavior by keeping track of the successful engine instances that got configured and for any subsequent errors invokes the task deletion API of the successful instances. If any of those fail, it just logs the warning and moves on with the others.

## How to test

A unit test was created for the `TasksService` and covers the success and failure cases.

### CODE REVIEW

File: src/main/java/com/rackspace/salus/event/manage/services/TasksService.java [hunk 1]
```diff
@@ -117,6 +139,10 @@ public EventEngineTask createTask(String tenantId, CreateTask in) {
     return eventEngineTaskRepository.save(eventEngineTask);
   }
 
+  private void rollbackTask(String kapacitorTaskId, List<EngineInstance> fromInstances) {
+
+  }
+
```
Comment (GeorgeJahad) [inline src/main/java/com/rackspace/salus/event/manage/services/TasksService.java]: is rollbackTask() needed?

Comment (itzg) [reply src/main/java/com/rackspace/salus/event/manage/services/TasksService.java]: Oops, no. I started that method before realizing `deleteTaskFromKapacitors` would work.

File: src/test/java/com/rackspace/salus/event/manage/services/TasksServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    verify(eventEnginePicker).pickAll();
+
+    verify(eventEngineTaskRepository).save(eventEngineTaskArgumentCaptor.capture());
+    assertThat(eventEngineTaskArgumentCaptor.getValue().getId())
+        .isEqualTo(taskId.getBaseId());
+
+    verifyNoMoreInteractions(eventEnginePicker, eventEngineTaskRepository, taskIdGenerator,
+        tickScriptBuilder
+    );
+  }
```
Comment (GeorgeJahad) [inline src/test/java/com/rackspace/salus/event/manage/services/TasksServiceTest.java]: should we also be verifying the mockKapacitorServer?:   https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/test/web/client/MockRestServiceServer.html#verify--

Comment (itzg) [reply src/test/java/com/rackspace/salus/event/manage/services/TasksServiceTest.java]: When I had bugs in the code/test it the `expect` calls already seem to be flagging the mismatches...I wonder if Spring Boot's unit test wrapper is calling verify. I'll investigate more.

**Comment by GeorgeJahad** [comment]
Is it worth adding a test for TasksService::deleteTask?

File: src/test/java/com/rackspace/salus/event/manage/services/TasksServiceTest.java (hunk 2 above)
Comment (itzg) [reply src/test/java/com/rackspace/salus/event/manage/services/TasksServiceTest.java]: ...using the debugger I confirmed the verify wasn't getting called already, so I'm adding these.

**Comment by itzg** [comment]
> Is it worth adding a test for TasksService::deleteTask?

As you can probably, I was adding the bare minimum of unit tests for the new code. But good question, since I'm already in here I'll cover the deleteTask also.

**Pull request merged**